File: generated-code/codecraft/rust/src/model/player_view.rs
use super::*;

#[derive(Clone, Debug)]
pub struct PlayerView {
    pub my_id: i32,
    pub map_size: i32,
    pub fog_of_war: bool,
    pub entity_properties: std::collections::HashMap<EntityType, EntityProperties>,
    pub max_tick_count: i32,
    pub max_pathfind_nodes: i32,
    pub current_tick: i32,
    pub players: Vec<Player>,
    pub entities: Vec<Entity>,
}

impl trans::Trans for PlayerView {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.my_id.write_to(writer)?;
        self.map_size.write_to(writer)?;
        self.fog_of_war.write_to(writer)?;
        self.entity_properties.write_to(writer)?;
        self.max_tick_count.write_to(writer)?;
        self.max_pathfind_nodes.write_to(writer)?;
        self.current_tick.write_to(writer)?;
        self.players.write_to(writer)?;
        self.entities.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let my_id: i32 = trans::Trans::read_from(reader)?;
        let map_size: i32 = trans::Trans::read_from(reader)?;
        let fog_of_war: bool = trans::Trans::read_from(reader)?;
        let entity_properties: std::collections::HashMap<EntityType, EntityProperties> = trans::Trans::read_from(reader)?;
        let max_tick_count: i32 = trans::Trans::read_from(reader)?;
        let max_pathfind_nodes: i32 = trans::Trans::read_from(reader)?;
        let current_tick: i32 = trans::Trans::read_from(reader)?;
        let players: Vec<Player> = trans::Trans::read_from(reader)?;
        let entities: Vec<Entity> = trans::Trans::read_from(reader)?;
        Ok(Self {
            my_id,
            map_size,
            fog_of_war,
            entity_properties,
            max_tick_count,
            max_pathfind_nodes,
            current_tick,
            players,
            entities,
        })
    }
}

File: generated-code/codecraft/haskell/FileReadWrite/src/Model/PlayerView.hs
module Model.PlayerView where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Data.Map
import Model.Entity (Entity)
import Model.EntityProperties (EntityProperties)
import Model.EntityType (EntityType)
import Model.Player (Player)

data PlayerView = PlayerView {
    myId :: Int32,
    mapSize :: Int32,
    fogOfWar :: Bool,
    entityProperties :: Map EntityType EntityProperties,
    maxTickCount :: Int32,
    maxPathfindNodes :: Int32,
    currentTick :: Int32,
    players :: [Player],
    entities :: [Entity] }
    deriving Show

instance Trans PlayerView where
    read = do
        myId <- Trans.read
        mapSize <- Trans.read
        fogOfWar <- Trans.read
        entityProperties <- Trans.read
        maxTickCount <- Trans.read
        maxPathfindNodes <- Trans.read
        currentTick <- Trans.read
        players <- Trans.read
        entities <- Trans.read
        return PlayerView {
            myId,
            mapSize,
            fogOfWar,
            entityProperties,
            maxTickCount,
            maxPathfindNodes,
            currentTick,
            players,
            entities }

    write PlayerView {
        myId,
        mapSize,
        fogOfWar,
        entityProperties,
        maxTickCount,
        maxPathfindNodes,
        currentTick,
        players,
        entities } = do
            Trans.write myId
            Trans.write mapSize
            Trans.write fogOfWar
            Trans.write entityProperties
            Trans.write maxTickCount
            Trans.write maxPathfindNodes
            Trans.write currentTick
            Trans.write players
            Trans.write entities